src/Application.php

```diff
@@ -232,41 +251,47 @@ public function notifyRouteResultObservers(Router\RouteResult $result)
      * the upshot is that you will not be notified if the service is invalid to
      * use as middleware until runtime.
      *
-     * Additionally, ensures that the route middleware is only ever registered
+     * Middleware may also be passed as an array; each item in the array must
+     * resolve to middleware eventually (i.e., callable or service name).
+     *
+     * Finally, ensures that the route middleware is only ever registered
      * once.
      *
-     * @param string|callable $path Either a URI path prefix, or middleware.
-     * @param null|string|callable $middleware Middleware
+     * @param string|array|callable $path Either a URI path prefix, or middleware.
+     * @param null|string|array|callable $middleware Middleware
      * @return self
      */
     public function pipe($path, $middleware = null)
     {
-        // Lazy-load middleware from the container when possible
-        $container = $this->container;
-        if (null === $middleware && is_string($path) && $container && $container->has($path)) {
-            $middleware = $this->marshalLazyMiddlewareService($path, $container);
-            $path       = '/';
-        } elseif (is_string($middleware)
-            && ! is_callable($middleware)
-            && $container
-            && $container->has($middleware)
+        if (null === $middleware) {
+            $middleware = $this->prepareMiddleware($path, $this->container);
+            $path = '/';
+        }
+
+        if (! is_callable($middleware)
+            && (is_string($middleware) || is_array($middleware))
         ) {
-            $middleware = $this->marshalLazyMiddlewareService($middleware, $container);
-        } elseif (null === $middleware && is_callable($path)) {
-            $middleware = $path;
-            $path       = '/';
+            $middleware = $this->prepareMiddleware($middleware, $this->container);
         }
 
         if ($middleware === [$this, 'routeMiddleware'] && $this->routeMiddlewareIsRegistered) {
             return $this;
         }
 
+        if ($middleware === [$this, 'dispatchMiddleware'] && $this->dispatchMiddlewareIsRegistered) {
```

Comment:
It's to prevent double-registration; most likely you would never want to double-route/double-dispatch, as that's expensive to do in the same request.

We can investigate making these exceptions instead, but that's out of scope for this patch; the behavior added for the dispatch middleware was done to mimic the existing behavior of registering the routing middleware.